controller/api/public/stat_summary.go

```diff
@@ -237,6 +257,80 @@ func (s *grpcServer) k8sResourceQuery(ctx context.Context, req *pb.StatSummaryRe
 	return resourceResult{res: &rsp, err: nil}
 }
 
+func (s *grpcServer) trafficSplitResourceQuery(ctx context.Context, req *pb.StatSummaryRequest) resourceResult {
+	requestedResource := req.GetSelector().GetResource()
+	objects, err := s.k8sAPI.GetObjects(requestedResource.Namespace, requestedResource.Type, requestedResource.Name)
+	if err != nil {
+		return resourceResult{res: nil, err: err}
+	}
+
+	tsBasicStats := make(map[tsKey]*pb.BasicStats)
+	rows := make([]*pb.StatTable_PodGroup_Row, 0)
+
+	for _, object := range objects {
+		ts, ok := object.(*v1alpha1.TrafficSplit)
+		if !ok {
+			return resourceResult{res: nil, err: errors.New("Could not cast to trafficsplit")}
+		}
+
+		backends := ts.Spec.Backends
+
+		tsStats := &trafficSplitStats{
+			namespace: ts.ObjectMeta.Namespace,
+			name:      ts.ObjectMeta.Name,
+			apex:      ts.Spec.Service,
+		}
+
+		if !req.SkipStats {
+			tsBasicStats, err = s.getTrafficSplitMetrics(ctx, req, tsStats, req.TimeWindow)
+			if err != nil {
+				return resourceResult{res: nil, err: err}
+			}
+		}
+
+		for _, backend := range backends {
+			name := fmt.Sprint(backend.Service)
+			weight := fmt.Sprint(backend.Weight.String())
+
+			currentLeaf := tsKey{
+				Namespace: tsStats.namespace,
+				Type:      k8s.TrafficSplit,
+				Name:      tsStats.name,
+				Apex:      tsStats.apex,
+				Leaf:      name,
+			}
+
+			trafficSplitStats := &pb.TrafficSplitStats{
+				Apex:   tsStats.apex,
+				Leaf:   name,
+				Weight: weight,
+			}
+
+			row := pb.StatTable_PodGroup_Row{
+				Resource: &pb.Resource{
+					Name:      tsStats.name,
+					Namespace: tsStats.namespace,
+					Type:      req.GetSelector().GetResource().GetType(),
+				},
+				TimeWindow: req.TimeWindow,
+				Stats:      tsBasicStats[currentLeaf],
```

Comment:
Ok this is super subtle, but I don't think we want this API to return rows for traffic splits that don't have any corresponding metrics in prometheus.

By way of example, in my `default` namespace, I'm running the books app, with an additional books service (`books-v2`). When I query for authorities, I see:

```
$ bin/linkerd stat au
NAME                                      MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
                                               -   100.00%   0.5rps           3ms           9ms          10ms          -
authors.default.svc.cluster.local:7001         -    77.09%   6.0rps           7ms          35ms          70ms          -
books-v2.default.svc.cluster.local:7002        -    75.00%   1.2rps          13ms         125ms         185ms          -
books.default.svc.cluster.local:7002           -    81.70%   6.5rps           9ms          86ms          97ms          -
webapp.default.svc.cluster.local:7000          -    78.70%   6.4rps          32ms          95ms         171ms          -
```

If I then add a `--from` flag, that filters the rows returned to only those that are called by the from resource. For instance:

```
$ bin/linkerd stat au --from deploy/webapp
NAME                                     MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
authors.default.svc.cluster.local:7001        -   100.00%   3.1rps          12ms          47ms          86ms          -
books.default.svc.cluster.local:7002          -    77.96%   7.0rps          12ms          83ms          97ms          -
```

Or:

```
$ bin/linkerd stat au --from deploy/books
NAME                                     MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
authors.default.svc.cluster.local:7001        -    46.60%   3.2rps           5ms          11ms          18ms          -
```

But that's not the behavior that we're seeing with traffic splits. If I query for all traffic splits in my default namespace, I see:

```
$ bin/linkerd stat ts
NAME          APEX    LEAF       WEIGHT   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
books-split   books   books        800m    79.63%   6.4rps          12ms          75ms         120ms
books-split   books   books-v2     200m    77.00%   1.7rps          20ms          33ms          39ms
```

If I add a `--to` flag, however, to filter to only those traffic splits that are hitting `svc/books-v2`, I see:

```
$ bin/linkerd stat ts --to svc/books-v2
NAME          APEX    LEAF       WEIGHT   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
books-split   books   books        800m         -        -             -             -             -
books-split   books   books-v2     200m    77.11%   1.4rps          10ms          62ms          92ms
```

Notice how the `books` row still shows up, even though it doesn't have any stats. For parity with how stat works for other resources, I think we should drop that row.

And the good news is that there's a very straightforward fix for this, by changing this func as follows:

```diff
diff --git a/controller/api/public/stat_summary.go b/controller/api/public/stat_summary.go
index 3763c2b9..afb3e077 100644
--- a/controller/api/public/stat_summary.go
+++ b/controller/api/public/stat_summary.go
@@ -310,6 +310,10 @@ func (s *grpcServer) trafficSplitResourceQuery(ctx context.Context, req *pb.Stat
 				Apex:      tsStats.apex,
 				Leaf:      name,
 			}
+			stats, ok := tsBasicStats[currentLeaf]
+			if !ok && !req.SkipStats {
+				continue
+			}
 
 			trafficSplitStats := &pb.TrafficSplitStats{
 				Apex:   tsStats.apex,
@@ -324,7 +328,7 @@ func (s *grpcServer) trafficSplitResourceQuery(ctx context.Context, req *pb.Stat
 					Type:      req.GetSelector().GetResource().GetType(),
 				},
 				TimeWindow: req.TimeWindow,
-				Stats:      tsBasicStats[currentLeaf],
+				Stats:      stats,
 				TsStats:    trafficSplitStats,
 			}
 			rows = append(rows, &row)
```

Reply:
Thanks for testing this out and thinking about this. You make a good point that there's not exact parity here between `ts` and other resources. 

My preference is to continue showing a row for each leaf within a trafficsplit even if there are no metrics returned, because in the case of traffic splitting, the *absence* of traffic is important information to validate that the split is working.

Also, if you are running `watch linkerd stat ts` as you apply weight changes within a split, showing a row for each leaf maintains a consistent display even if a leaf service you are phasing out goes down to 0 (or up from 0), which is more readable than a row popping in or out of view. WDYT?

You're right that we'll need to be specific in the descriptions about what --from and --to means in the case of `linkerd stat ts`. This is a very wordy attempt that needs to be more concise:

```
linkerd stat ts —from deploy/authors
Show all trafficsplit leaf services in the default namespace, and metrics for any traffic coming to those services from the authors deployment.
```

Reply:
Personally, I also think it's important to show the traffic split branches with no metrics even though this is inconsistent with how other resources behave.

Reply:
Ok, this all sounds good to me. I just wanted to clarify that if a traffic split has ever received any traffic that matches the query pattern, then it will still be displayed, even if it currently is not receiving any traffic. My proposed change would only exclude rows that have never seen a single request matching the query pattern.

But I agree that the nature of trafficsplits is different enough from other resources to warrant showing all branches in the stat output.